curvefs/src/client/s3/client_s3_adaptor.cpp

```diff
@@ -0,0 +1,623 @@
+        if (version == tmp->version()) {
+            if (offset == tmp->offset() + tmp->len()) {
+                len += tmp->len();
+                tmp->set_len(len);
+                offset = tmp->offset();
+                merge = true;
+            }
+        }
+    }
+
+    if (!merge) {
+        tmp = s3ChunkInfoList->add_s3chunks();
+        tmp->set_chunkid(chunkId);
+        tmp->set_version(version);
+        tmp->set_offset(offset);
+        tmp->set_len(len);
+        tmp->set_size(len);
+    }
+
+    return;
+}
+
+int S3ClientAdaptor::GetChunkId(Inode *inode, uint64_t index,
+                                uint64_t *chunkId) {
+    S3ChunkInfo tmp;
+    uint64_t tmpIndex;
+    CURVEFS_ERROR retCode = CURVEFS_ERROR::FAILED;
+    const S3ChunkInfoList &s3chunkInfoList = inode->s3chunkinfolist();
+    if (inode->length() == 0) {
+        retCode = AllocS3ChunkId(inode->fsid(), chunkId);
+        if (retCode != CURVEFS_ERROR::OK) {
+            return -1;
+        }
+        LOG(INFO) << "GetChunkId():chunkid:" << *chunkId;
+        return 0;
+    }
+
+    for (int i = 0; i < s3chunkInfoList.s3chunks_size(); ++i) {
+        tmp = s3chunkInfoList.s3chunks(i);
+        tmpIndex = tmp.offset() / chunkSize_;
+        if (tmpIndex < index) {
+            continue;
+        } else if (tmpIndex == index) {
+            *chunkId = tmp.chunkid();
+            return 0;
+        }
+    }
+
+    retCode = AllocS3ChunkId(inode->fsid(), chunkId);
+    if (retCode != CURVEFS_ERROR::OK) {
+        return -1;
+    }
+    LOG(INFO) << "GetChunkId():chunkid:" << *chunkId;
+    return 0;
+}
+
+int S3ClientAdaptor::UpdateInodeS3Version(Inode *inode, uint64_t *version) {
+    brpc::Channel channel;
+
+    if (channel.Init(metaServerEps_.c_str(), NULL) != 0) {
+        LOG(ERROR) << "Fail to init channel to meta Server"
+        << " for update inode version: " << metaServerEps_;
+        return -1;
+    }
+
+    brpc::Controller* cntl = new brpc::Controller();
+    UpdateInodeS3VersionRequest request;
+    UpdateInodeS3VersionResponse response;
+
+    request.set_inodeid(inode->inodeid());
+    request.set_fsid(inode->fsid());
+    curvefs::metaserver::MetaServerService_Stub stub(&channel);
+
+    stub.UpdateInodeS3Version(cntl, &request, &response, NULL);
+
+    if (cntl->Failed()) {
+        LOG(WARNING) << "Update inode s3 version Failed, errorcode = "
+                     << cntl->ErrorCode()
+                     << ", error content:" << cntl->ErrorText()
+                     << ", log id = " << cntl->log_id();
+        return cntl->ErrorCode();
+    }
+
+    curvefs::metaserver::MetaStatusCode ssCode = response.statuscode();
+    if (ssCode != curvefs::metaserver::MetaStatusCode::OK) {
+        LOG(WARNING) << "update inode s3 version response Failed, retCode = "
+                     << ssCode;
+        return -1;
+    }
+
+    *version = response.version();
+    return 0;
+}
+
+CURVEFS_ERROR S3ClientAdaptor::AllocS3ChunkId(uint32_t fsId,
```

Comment:
why some function use CURVEFS_ERROR as return type，and others use int？

Reply:
this problem must be modified later